### ISSUE TITLE
[improvement] react/no-unstable-nested-components & false negatives

### ISSUE THREAD

**Issue by xgbuils** [jsx-eslint/eslint-plugin-react#3621]
I've seen that this rules has a lot of false negatives.

A component is allowed to be defined inside other component if is created by a hook or a HOC. For example:

``` JavaScript
function ParentComponent() {
  const InsideDefinedFunctionComponent = withHappiness(OutsideDefinedFunctionComponent);
  return (
    <div>
      <InsideDefinedFunctionComponent />
    </div>
  );
}
```

Sorry [WIP]
      `

**Comment by ljharb** [comment]
I don't think that should be allowed - in that example, you should define the component at the module level, so it's only done once.

In particular, using an HOC *and* a hook makes no sense - one reason hooks were invented was to avoid the need for HOC patterns.

**Comment by xgbuils** [comment]
I'm gonna close this ticket and create a new one. I pressed Ctrl+Enter involuntarily.

**Comment by ljharb** [comment]
Please just edit and reopen this one.
